iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp

```diff
@@ -211,159 +210,22 @@ struct FoldAsNoOp final : public OpConversionPattern<OpTy> {
   }
 };
 
-/// Base class for lowering to SPIR-V cooperative matrix ops.
-template <typename SourceOp>
-class CoopMatOpLowering : public OpConversionPattern<SourceOp> {
- public:
-  CoopMatOpLowering(MLIRContext *context, SPIRVTypeConverter &converter,
-                    // Dedicated extensions are typically faster; so give it a
-                    // higher benefit so it prevails by default.
-                    PatternBenefit benefit = 5)
-      : OpConversionPattern<SourceOp>(context, benefit), converter(converter) {}
-
- protected:
-  // TODO: We explicitly keep a reference of the type converter instead of
-  // passing it to OpConversionPattern during construction. This effectively
-  // bypasses the dialect conversion framework's automation over type
-  // conversion. This is needed for now because upstream SPIRVTypeConverter does
-  // not support cooperative matrix well yet so the framework won't know how to
-  // generate cooperative matrix. We are manually constructing the cooperative
-  // matrix in patterns. This should be fixed when we upstream all cooperative
-  // matrix related code.
-  SPIRVTypeConverter &converter;
-};
-
-/// Convert subgroup level vector transfert to SPIR-V cooperative
-/// matrix load/store if those are supported.
-/// TODO(thomasraoux): Move to MLIR core once this is stable.
-template <typename OpTy>
-class TransferToCoopMatLoadStore final : public CoopMatOpLowering<OpTy> {
- public:
-  TransferToCoopMatLoadStore(
-      MLIRContext *context, SPIRVTypeConverter &converter,
-      const CooperativeMatrixAnalysis &cooperativeMatrixAnalysis)
-      : CoopMatOpLowering<OpTy>(context, converter),
-        cooperativeMatrixAnalysis(cooperativeMatrixAnalysis) {}
-
+/// Removes unrealized_conversion_cast ops introduced during progressive
+/// lowering when possible.
+struct RemoveIdentityConversionCast final
+    : public OpConversionPattern<UnrealizedConversionCastOp> {
+  using OpConversionPattern::OpConversionPattern;
   LogicalResult matchAndRewrite(
-      OpTy op, ArrayRef<Value> operands,
+      UnrealizedConversionCastOp op, ArrayRef<Value> operands,
```

Comment:
I don't understand why it's scary? This is just a simple canonicalization pattern, right? Unrealized conversion casts are expected to be the "bridges" between progressive lowerings and it should go away at the very last step when we convert everything to the final target, which is in this pass. Unrealized conversion casts here were inserted by dialect conversion framework to "cast" `hal.interface.binding.subspan` ops into its SPIR-V type, which won't be materialized until in this pass where we convert `hal.interface.binding.subspan` into SPIR-V global variables.

Reply:
Ill approve this PR since it contains things better which is good, but to me `UnrealizedConversionCast` seems like a catch all, and an indication that something "wasnt handled". It is very easy for a lot of things to end up there, and very easy for something to work even though it was not suppsoed to work. Let me ask a different question, if the lowering of `vector.contract` to `spv.CoooperativeMatrix*` was done in this pass and you had the `hal.interface.binding.subspan` to SPIR-V type patterns, these instructions wouldnt exist right? Its helping you break the lowering of vector.contract to cop-operative matrix instructions and general lowering to SPIR-V, but the real issue is that the SPIR-V lowering to co-operative matrix multiply instructions need to be fixed in core to handle the different types used for these ops.

> > I am fine with moving this around, but we should just move this to core. The issue here is that the co-operative matrix has its own types and load-store instructions. So the vector transfer-reads and transfer-writes become problematic. THe best way to handle this is to update the Vector to Co-operative matrix lowering in core to not require the pre-analysis pass done here. For this we are first missing a "spv.cast" operation that converts from the default `spv.array` type to the co-operative matrix type.
> > 
> > 1. Lower the `vector.contract` using `spv.cast` -> `spv.cooperativeMatrixMulOp` -> `spv.cast`.
> > 2. Then you have a `spv.load` -> `spv.cast` that should be converted to `spv.loadcooperativematrix` (and same for the store).
> > 
> > This then requires no state and this pass can be dropped. Would be better for SPIR-V lowering in general.
> > This refactoring by itself is good, but deleting this pass completely is even better (we have code cleanup as part of our OKRs this quarter)
> 
> Yes right. The existing analysis is hacky. @ThomasRaoux is pushing for a more generic cooperative matrix ish representation in gpu dialect so that we can share between cuda and spirv land. After that's landed we can go from vector to gpu counterparts and then convert to spirv will be simple. Expected this part to be simplified in the future; right now doing this to unblock deprecating the buffer path.

@ThomasRaoux we need to talk about this. Use of GPU dialect might just displace the issue. The real fix needs to be someway in which SPIR-V itself can handle the special type for co-operative matrix multiplies.

Reply:
Thanks. There are multiple issues; type conversion being one of them. There certainly should be direct support in the SPIR-V type converter upstream to convert into cooperative matrix types (essentially the additional conversion rules I attached in this pass, with proper capability check, etc.). Having all of those properly implemented takes time. I mentioned Thomas' plans just to show that he is working on pushing for better solutions already, so we can expect this to be improved. Not saying going gpu is directly addressing all the issues; that was more for sharing with cuda side. Hope this clarifies. :)

Reply:
> @ThomasRaoux we need to talk about this. Use of GPU dialect might just displace the issue. The real fix needs to be someway in which SPIR-V itself can handle the special type for co-operative matrix multiplies.

Sure, let's discuss this. I think the GPU dialect should address most of the problem since it will give us an op working on memref analogue to std.load and we can support the new type in the SPIR-V type converter.